### PR TITLE
Fixes launchbar icons looking blurry on hidpi screens

### PR DESCRIPTION
**Pull request by mbooth101**
This is a problem originally reported in https://bugs.eclipse.org/bugs/show_bug.cgi?id=569132

The launchbar way of compositing images for the buttons was preventing the use of higher quality icons on hidpi dislays, making the launchbar buttons look blurry or pixelated compared to other toolbar buttons.

Simply removing the border image compositing, as @akurtakov suggested in his original change, made the buttons lose their distinctive look, which was undesirable.

However, I believe we get a decent look by instead using a custom widget on which we can directly draw a button border plus the hidpi version of the icon. This looks way better IMO because we can give the button widgets the same matching vertical dimension and widget border style as the other launchbar widgets, giving a more uniform look and feel.

Also included in the change is an attempt at a 2x hammer build button icon -- however _I am not a pixel artist_ and while my new icon is not perfect, it should still look better than upsampling the small icon. (My artistic talent is best described as "far from good, but good from afar" :smiley: )

The improvement is most noticable when using the dark theme because the "hard-coded" border image was obviously not theme aware. Here's some side-by-side comparisons:

### Fedora HiDPI Before:
fedora-hidpi-old

### Fedora HiDPI After:
fedora-hidpi-new

### Fedora HiDPI Dark Before:
fedora-hidpi-dark-old

### Fedora HiDPI Dark After:
fedora-hidpi-dark-new

### Windows 10 HiDPI Before:
windows-hidpi-old

### Windows 10 HiDPI After:
windows-hidpi-new

### Ubuntu Low-DPI Before:
ubuntu-lodpi-old

### Ubuntu Low-DPI After:
ubuntu-lodpi-new

### CODE REVIEW

**Comment by mbooth101** [comment]
Test failures look unrelated.

@jonahgraham WDYT? Sorry there's no Mac screenshot, I don't have access to one.

**Review by jonahgraham — CHANGES_REQUESTED** [review]
This looks very good, a couple of nits to look at though please.

File: launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+			} else if (hover) {
+				e.gc.setBackground(getDisplay().getSystemColor(SWT.COLOR_WIDGET_HIGHLIGHT_SHADOW));
+			} else {
+				e.gc.setBackground(getBackground());
+			}
+			int arc = 3;
+			Point size = getSize();
+			e.gc.fillRoundRectangle(0, 0, size.x - 1, size.y - 1, arc, arc);
+			if (image != null) {
+				e.gc.drawImage(image, 9, 5);
```
[the remark this answers is not in the document]
Comment (mbooth101) [reply launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java]: Changed to 10,5

File: launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			hover = false;
+			redraw();
+		}));
+		addMouseListener(MouseListener.mouseDownAdapter(e -> {
+			pressed = true;
+			redraw();
+		}));
+		addMouseListener(MouseListener.mouseUpAdapter(e -> {
+			pressed = false;
+			notifyListeners(SWT.Selection, null);
```
[the remark this answers is not in the document]
Comment (mbooth101) [reply launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java]: Should be fixed now -- we don't get mouseExit whilst the mouse button is pressed, so added a mouseMove listener to accomodate this case. The behaviour should now be the same as a standard button

**Comment by mbooth101** [comment]
Thanks for the review Jonah, hopefully the latest patchset addresses the nits :-)

File: launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			}
+			e.gc.setForeground(getDisplay().getSystemColor(SWT.COLOR_WIDGET_NORMAL_SHADOW));
+			e.gc.drawRoundRectangle(0, 0, size.x - 1, size.y - 1, arc, arc);
+		});
+
+		addMouseMoveListener(e -> {
+			// We don't get enter and exit events whilst a mouse button is down i.e., during click-drag,
+			// so we also need to track the hover state here because during a click-drag we still get
+			// move events even when the pointer is not over the control
+			if (e.getSource() instanceof Control button) {
```
Comment (jonahgraham) [inline launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java:61]: 🎉 for use of pattern matching for `instanceof` - we'll get this ~20 year old code modernized yet!

**Review by jonahgraham — APPROVED** [review]

**Comment by jonahgraham** [comment]
> Thanks for the review Jonah, hopefully the latest patchset addresses the nits :-)

Thanks - it looks good on Linux. I haven't run on Windows/Mac.

File: launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java (hunk 3 above)
Comment (mbooth101) [reply launchbar/org.eclipse.launchbar.ui.controls/src/org/eclipse/launchbar/ui/controls/internal/CLaunchButton.java:61]: I never liked type-casting, so I :heart: this syntax :-)

**Pull request merged**